xdelta3-dir-patcher

```diff
@@ -21,6 +49,355 @@ class AttributeDict(dict):
+        else:
+            makedirs(target_path, exist_ok = True)
+
+            # TODO: Test me
+            # Ensure that permissions/ids are transferred along to the target
+            copymode(source_path, target_path)
+            copystat(source_path, target_path)
+
+            try:
+                uid = stat(source_path).st_uid
+                gid = stat(source_path).st_gid
+
+                chown(target_path,uid,gid)
+            except PermissionError as pe:
+                # XXX: We can't copy the uid/gid unless we're run as root
+                #      which means that in most cases we can't get that 
+                #      included in the diffs unless we add direct linkage
+                #      between archive<->archive to copy these values
+
+                # print('WARNING! Could not change uid/gid of', target_path)
+                pass
+
+    def create(self, base_dir):
+        if path.isdir(self.path):
+            raise Exception('Error! Archive already present!')
+
+        copytree(base_dir, self.path,
+                 ignore_dangling_symlinks = False)
+
+class XDelta3TarImpl(XDelta3AbstractArchiveImpl):
+    TAR_FORMAT = 'gz'
+
+    def __init__(self, archive_path, for_writing = False):
+        super().__init__()
+
+        self.items = None
+        flags = 'r:*'
+
+        if for_writing:
+            if path.isfile(archive_path):
+                raise Exception('Error! Archive already present!')
+
+            flags = 'w:%s' % self.TAR_FORMAT
+
+        self.archive_object = tarfile.open(archive_path, flags)
+
+    def _close_archive(self):
+        self.archive_object.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._close_archive()
+
+    def close(self):
+        self._close_archive()
+
+    @staticmethod
+    def can_open(archive):
+        return path.isfile(archive) and tarfile.is_tarfile(archive)
+
+    def list_files(self):
+        if self.items:
+            return self.items
+
+        items = []
+
+        print('Tar: Gathering filelist')
+        members = self.archive_object.getmembers()
+
+        # Lookup all member objects
+        for member in members:
+            # For directories, add a separator to the end
+            suffix = path.sep if member.isdir() else ''
+
+            items.append('%s%s' % (member.name, suffix))
+
+        # Append any directories that weren't part of the archive
+        for item in items:
+            dir_name = path.dirname(item) + path.sep
+            if len(dir_name) > 1 and not dir_name in items:
+                print('Post-adding %s to members' % dir_name)
+                items.append(dir_name)
+
+        print('Tar: Gathering completed')
+
+        self.items = items
+
+        return self.items
+
+    def expand(self, filename, extraction_path):
+        member = None
+        try:
+            member = self.archive_object.getmember(filename)
```

Comment:
@sgnn7 I think this is the quintessential reason to cache the getmembers output. You already got all this info before, and then you might call getmember a 3rd time below.